### PR TITLE
Treat feature example sections the same as scenarios

### PR DESCRIPTION
**Pull request by rbcasperson**
If all example sections are at the same level, it can cause an odd case
where a feature example comes before any scenario. In that case,
all examples tables' section headers will be larger than scenario
section headers. So this puts feature examples at the same level as
scenarios, which matches the indentation level in the feature file.

Here's that odd case:
Screen Shot 2020-05-18 at 4 21 33 PM

^ Notice that `Examples` is larger than `Scenario Outline` and `Scenario`. This all has to do with rST caring only about the order of section underlines rather than what they are. Since we treat all examples tables the same, and examples comes first in this case, it thinks Examples are a higher level.

The reality is that feature examples tables are one level below the feature, the same as all scenarios and scenario outlines. This PR changes the logic to do just that. Here is what it looks like fixed:
Screen Shot 2020-05-18 at 4 24 34 PM

^ Background, Feature Examples, and Scenarios are all on the same level (same size), and scenario examples are a subsection of scenarios.

### CODE REVIEW

File: tests/rst_output_pytest.py [hunk 1]
```diff
@@ -267,7 +267,7 @@
     ":gherkin-scenario-keyword:`Scenario:` :gherkin-scenario-content:`Test a Scenario`\n",
     "---------------------------------------------------------------------------------\n\n",
     ".. pull-quote::\n\n",
-    "    :gherkin-tag-keyword:`Tagged:` quarantined, JIRA-1234\n\n",
+    "    :gherkin-tag-keyword:`Tagged:` JIRA-1234, quarantined\n\n",
```
Comment (rbcasperson) [inline tests/rst_output_pytest.py:270]: This was an unexpected change. I can't currently explain why the order changed, but the tests keep passing over and over, so I'm sure it isn't random (if it is, I should play the lottery getting a coin to land on heads that many times in a row!)

File: sphinx_gherkindoc/writer.py [hunk 2]
```diff
@@ -315,14 +317,24 @@ def steps(steps: List[behave.model.Step], step_format: str = "{}") -> None:
...
         example_source_parent: Optional[behave.model.Feature] = None,
     ) -> None:
         tag_sources = [example_source]
         if example_source_parent:
             tag_sources.append(example_source_parent)
         for example in getattr(example_source, "examples", []):
-            section(3, example)
+            if isinstance(example_source, pytest_bdd_models.Feature):
+                # Set feature examples tables at the same section level as scenarios.
+                section_level = 2
```
Comment (dgou) [inline sphinx_gherkindoc/writer.py:333]: I wish there were a cleaner way to do this `if` but nothing is coming immediately to mind. RR.

File: tests/rst_output_pytest.py (hunk 1 above)
Comment (dgou) [reply tests/rst_output_pytest.py:270]: Since we expected to see the justification/Ticket tags after the non-operational status tags, this is concerning.
I agree, I don't see anything in this PR that would account for the change though.

File: sphinx_gherkindoc/writer.py [hunk 3]
```diff
@@ -315,14 +317,24 @@ def steps(steps: List[behave.model.Step], step_format: str = "{}") -> None:
                 text(step.text)
 
     def examples(
-        example_source: Union[behave.model.Scenario, behave.model.Feature],
+        example_source: Union[
+            behave_models.Scenario,
+            pytest_bdd_models.Scenario,
+            pytest_bdd_models.Feature,
```
Comment (dgou) [inline sphinx_gherkindoc/writer.py:323]: This is surprising, I expected to see `pytest_bdd_models.Feature` added to the `Union` but I did not expect to see `behave_models.Feature` removed...

File: tests/rst_output_pytest.py (hunk 1 above)
Comment (dgou) [reply tests/rst_output_pytest.py:270]: I checked and the behave output files always have the quarantined tag first.
But this isn't the only place where the pytest tags are out of order.
While this bugs me, so long as it stays pytest only, I am willing to defer the urgency of fixing it, in this PR, to those directly affected.

File: sphinx_gherkindoc/writer.py (hunk 3 above)
Comment (dgou) [reply sphinx_gherkindoc/writer.py:323]: I'd like to understand this change before signing off.

Comment (rbcasperson) [reply sphinx_gherkindoc/writer.py:323]: So the short of it is that the current type hints are actually off. Since we define our own models for each parser ourselves, the original `behave.model.Scenario, behave.model.Feature` actually would never be the type of `example_source`. Since it is related to these changes (the isinstance check), I made it correct here. `behave_models.Feature` can never be passed in here because it never can have a `.feature` value. The line that calls this checks first `if feature.examples:`.

I'll have a FF to clean up all of the type hints in this file to use the correct `behave_models` I added instead of the `behave.model` values, which are now inaccurate. You'll notice that no where else in the type hints are the pytest_bdd models mentioned, so they need to be everywhere really. I'll try to organize it concisely and clearly.

**Comment by dgou** [comment]
Ran this PR against my work repo, no changes there, so one question (per above) and I can sign off.

File: sphinx_gherkindoc/writer.py (hunk 2 above)
Comment (rbcasperson) [reply sphinx_gherkindoc/writer.py:333]: Yeah. I though about having constants like `FEATURE_LEVEL = 1` and `EXAMPLES_LEVEL = 2`, but I don't think those are really any more helpful than the plain numbers. Not sure if that's where your thoughts were heading. 

Otherwise, yeah I think `isinstance` is probably the best option. We only change the behavior if the source is of a certain type. We could pass in the level as a param and control it in the logic that calls this function, but it'd be the same `if` check.

File: tests/rst_output_pytest.py (hunk 1 above)
Comment (rbcasperson) [reply tests/rst_output_pytest.py:270]: Yes that's a good point. I'll try to take a closer look to figure this out. Maybe @bradsbrown has a hunch.

Comment (dgou) [reply tests/rst_output_pytest.py:270]: "Having slept on it", my "guess" is that the pytest bdd parser does not consider the order of the tags to be important and is not using an order preserving data type to accumulate them. Given that, IIRC, it doesn't even support tags on multiple lines, I would not be surprised to see something like-ish `set(tag.strip() for tag in tag_line.split())`
Why that would suddenly change though is more curious than why it was never ordered in the first place. If the underlying mechanism is not ordered, then doing anything about it is out of scope for this PR.

tl;dr this only affects pytest so I am willing to sign off because behave is still working properly.

File: sphinx_gherkindoc/writer.py (hunk 2 above)
Comment (dgou) [reply sphinx_gherkindoc/writer.py:333]: Having slept on this too, I am surprised I ever signed off on having hard coded constants in the first place.
At least having `section_level = FEATURE_EXAMPLE_LEVEL` and `section_level = SCENARIO_EXAMPLE_LEVEL` would help clarify some, but there are cleanups for another time.

File: sphinx_gherkindoc/writer.py (hunk 3 above)
Comment (dgou) [reply sphinx_gherkindoc/writer.py:323]: Ok, so basically the type hinting is wrogn and will be fixed in another PR.

File: tests/rst_output_pytest.py (hunk 1 above)
Comment (bradsbrown) [reply tests/rst_output_pytest.py:270]: That's close enough to verbatim [how pytest-bdd gets tags](https://github.com/pytest-dev/pytest-bdd/blob/master/pytest_bdd/feature.py#L142).

Comment (bradsbrown) [reply tests/rst_output_pytest.py:270]: For the record, I'm guessing the expected result file changed because you had other updates to make in it, but the reason the tests consistently pass regardless of tag order is that we [intentionally compare tags as a set](https://github.com/jolly-good-toolbelt/sphinx_gherkindoc/blob/master/tests/test_pytest_writer.py#L46) between the actual and expected so that we can be as apathetic about tag order as `pytest-bdd` has chosen to be.

Comment (rbcasperson) [reply tests/rst_output_pytest.py:270]: That explains it! So this change really has no effect one way or another. It might be worth noting somewhere that tag order is not guaranteed when using pytest-bdd parser, but right now no one we know is relying on that. If there's an obvious place to put it, I'll drop it in here, but otherwise I'll FF.

File: sphinx_gherkindoc/writer.py (hunk 3 above)
Comment (bradsbrown) [reply sphinx_gherkindoc/writer.py:323]: Uf, I hate that I (and `mypy`?) missed this in the midst of the larger work.

File: sphinx_gherkindoc/writer.py (hunk 2 above)
Comment (brolewis) [reply sphinx_gherkindoc/writer.py:333]: @dgou I'm pretty sure all those hard-coded constants were in the original script I wrote back at Rackspace and you _might_ not have been involved in the project yet. Not saying we can't make it better, but to at least let you and Ryan off the hook for now 😁
